src/client/Microsoft.Identity.Client/OAuth2/TokenClient.cs

```diff
@@ -161,6 +161,54 @@ private void AddBodyParamsAndHeaders(IDictionary<string, string> additionalBodyP
             {
                 _oAuth2Client.AddHeader(PKeyAuthConstants.DeviceAuthHeaderName, PKeyAuthConstants.DeviceAuthHeaderValue);
             }
+
+            AddCCSRoutingInformation(additionalBodyParameters);
```

Comment:
I'm in 2 minds about having this functionality here. On one side, the code is pretty compact. On the other, it violates the single reponsability principle, i.e. the the tokenclient class neesd to format requests to the token endpoint. It shouldn't need to know how to extract CCS information. The request classes themselves are a better place for this kind of knowledge, smth like; 


```csharp
//in RequestBase.cs define a GetCCSHeader() which every request that can provide this must implement
protected string GetCCSHeader() { return null; }

// modify SendTokenRequestAsync to inject this extra header into the tokenClient it creates.

```

Reply:
It's also a slightly more performant solution, as `client_credentials` request will simply return `null` instead of going though all those if conditions and string comparisons.

Reply:
Since all of the request classes simply call the base SendTokenRequestAsync and the base class has all of the data necessary to the header most of the time I only implemented the method in the classes that could not provide any data or if the data was not available to the requestBase